File: algds/leetcode/medium/house-robber/py/rob_v1.py
#
# LeetCode: House Robber :: Medium
#
# tags: dynamic-programming
#

##
# • T.C: O(n)
# • S.C: O(1)
#
def rob(nums):
  max1, max2 = 0, 0

  for n in nums:
    tmp = max(max1, n + max2)
    max2 = max1
    max1 = tmp

  return max1



print(rob([]))
#=> 0

print(rob([7]))
#=> 7

print(rob([3, 1]))
#=> 3

print(rob([1, 2, 5]))
#=> 6

print(rob([1, 2, 3, 1]))
#=> 4

print(rob([2, 7, 9, 3, 1]))
#=> 12

print(rob([7, 7, 8, 7]))
#=> 15

print(rob([3, 5, 1, 9, 9]))
#=> 14

File: algds/leetcode/medium/house-robber/C/rob_v1.c
#include <stdio.h>

#define MAX(x, y) (((x) > (y)) ? (x) : (y))

int rob(int* nums, int len) {
  int tmp;
  int max1 = 0;
  int max2 = 0;

  for (int i = 0; i < len; ++i) {
    tmp = MAX(max1, *(nums + i) + max2);
    max2 = max1;
    max1 = tmp;
  }

  return max1;
}

int main(void) {
  printf("Hey\n");

  int nums1[] = { 7 };
  int nums2[] = { 3, 7 };
  int nums3[] = { 1, 2, 5 };
  int nums4[] = { 1, 2, 3, 1 };
  int nums5[] = { 2, 7, 9, 3, 1 };

  printf("%d\n", rob(nums1, 1));
  //=> 7

  printf("%d\n", rob(nums2, 2));
  //=> 7

  printf("%d\n", rob(nums3, 3));
  //=> 6

  printf("%d\n", rob(nums4, 4));
  //=> 4

  printf("%d\n", rob(nums5, 5));
  //=> 12

  return 0;
}
